drivers/wireless/lpwan/rn2xx3/rn2xx3.c

```diff
@@ -0,0 +1,1508 @@
+
+static const struct file_operations g_rn2xx3fops =
+{
+  .open = NULL,
+  .close = NULL,
+  .read = rn2xx3_read,
+  .write = rn2xx3_write,
+  .ioctl = rn2xx3_ioctl,
+};
+
+/* Modulation types */
+
+static const char *MODULATIONS[] =
+{
+  "lora", /* RN2XX3_MOD_LORA */
+  "fsk",  /* RN2XX3_MOD_FSK */
+};
+
+/* Max packet sizes for each modulation type. */
+
+static const uint8_t MAX_PKTSIZE[] =
+{
+  255, /* RN2XX3_MOD_LORA */
+  64,  /* RN2XX3_MOD_FSK */
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: radio_flush
+ *
+ * Description:
+ *   Flushes all characters in the receive buffer from the RN2xx3 radio.
+ *   Returns the number of bytes read.
+ *
+ ****************************************************************************/
+
+static int radio_flush(FAR struct rn2xx3_dev_s *priv)
+{
+  int err;
+  int to_read;
+  char buf[10];
+
+  err = file_ioctl(&priv->uart, FIONREAD, &to_read);
+  if (err < 0)
+    {
+      return err;
+    }
+
+  while (to_read > 0)
+    {
+      err = file_read(&priv->uart, buf, sizeof(buf));
+      if (err < 0)
+        {
+          return err;
+        }
+
+      to_read -= err;
+    }
+
+  return 0;
+}
+
+/****************************************************************************
+ * Name: read_line
+ *
+ * Description:
+ *   Reads a whole line of response from the RN2xx3 radio. Returns the number
+ *   of bytes read.
+ *
+ ****************************************************************************/
+
+static ssize_t read_line(FAR struct rn2xx3_dev_s *priv, FAR char *buf,
+                         size_t nbytes)
+{
+  ssize_t length = 0;
+  ssize_t i = 0;
+
+  for (; i < nbytes; )
+    {
+      length = file_read(&priv->uart, &buf[i], nbytes - i);
+
+      if (length < 0)
+        {
+          return length;
+        }
+
+      i += length;
+
+      if (buf[i] == '\n' || (buf[i - 1] == '\n' && i > 0))
+        {
+          break;
+        }
+    }
+
+  return i + 1; /* Number of bytes read */
+}
+
+/****************************************************************************
+ * Name: mac_pause
+ *
+ * Description:
+ *   Pauses the MAC layer. Required before every transmission/receive.
+ *
+ ****************************************************************************/
+
+static int mac_pause(FAR struct rn2xx3_dev_s *priv)
+{
+  ssize_t length = 0;
+  char response[30];
+
+  /* Issue pause command */
+
+  length =
+      file_write(&priv->uart, "mac pause\r\n", sizeof("mac pause\r\n") - 1);
+  if (length < 0)
+    {
+      return length;
+    }
+
+  /* Wait for response of watchdog timer timeout */
+
+  length = read_line(priv, response, sizeof(response));
+  if (length < 0)
+    {
+      return length;
+    }
+
+  response[length] = '\0'; /* Avoid strstr overrun */
```

Comment:
You write out of bound when `read_line` returns `sizeof(response)`. If `read_line` returns valid NULL terminated string, just do `response[sizeof(response) - 1]`. Otherwise sanitize length to not write out of bounds.